pkg/ipcache/metadata.go

```diff
@@ -327,43 +417,64 @@ func RemoveLabelsFromIPs(
 
 		triggerer.TriggerPolicyUpdates(false, "kube-apiserver identity updated by removal")
 	}
+	for ip, pair := range toReplace {
+		hIP, key := IPIdentityCache.getHostIPCache(ip)
+		meta := IPIdentityCache.getK8sMetadata(ip)
+		IPIdentityCache.deleteLocked(ip, pair.old.Source) // remove old entry and replace
+		if _, err := IPIdentityCache.upsertLocked(ip, hIP, key, meta, Identity{
+			ID:     pair.new.ID,
+			Source: pair.new.Source,
+		}); err != nil {
+			log.WithError(err).WithFields(logrus.Fields{
+				logfields.IPAddr:   ip,
+				logfields.Identity: pair.new,
+			}).Error("Failed to replace ipcache entry with new identity after label removal. Traffic may be disrupted.")
+		}
+	}
+}
+
+type idPair struct {
+	old, new Identity
 }
 
-// RemoveLabels removes the given labels association with the given prefix. The
-// leftover labels are returned, if any.
+// removeLabels removes the given labels association with the given prefix. The
+// leftover labels are returned, if any, and a boolean indicating whether the
+// caller must generate a new identity with the leftover labels. If the caller
+// must do so, then the following must occur *in order* to avoid drops:
+//   1) policy recalculation must be triggered
+//   2) new identity must have a new entry upserted into the IPCache
 //
 // Identities are deallocated and their subequent entry in the IPCache is
 // removed if the prefix is no longer associated with any labels.
 //
+// This function assumes that the IDMD and the IPIdentityCache locks are taken!
+//
 // It is the responsibility of the caller to trigger policy recalculation after
 // calling this function.
-func RemoveLabels(prefix string, lbls labels.Labels, src source.Source) labels.Labels {
-	idMDMU.Lock()
-	defer idMDMU.Unlock()
+func removeLabels(prefix string, lbls labels.Labels, src source.Source) (labels.Labels, bool) {
+	var needNewIdentity bool
 
 	l, ok := identityMetadata[prefix]
 	if !ok {
-		return nil
+		return nil, needNewIdentity
 	}
 
 	l = l.Remove(lbls)
 	if len(l) != 0 { // Labels left over, do not deallocate
 		identityMetadata[prefix] = l
-		return l
+		return l, needNewIdentity
```

Comment:
If there are labels left over, then how do we know that there's an identity associated with the new set of labels? Don't we need to at least check whether we should allocate an identity for that new set of labels?

Reply:
Yes that's handled at the caller when `needNewIdentity=true`. Am I missing something?

Reply:
:thinking: Good point, I think this is an oversight. Let me do some thinking as to why we even have `Labels left over, do not deallocate`...

Reply:
Should be addressed in latest commit now